File: CashuWallet/Core/PaymentRequestDecoder.swift
import Foundation
import Cdk

/// Display-safe summary of a NUT-18/NUT-26 Cashu payment request.
struct CashuPaymentRequestSummary: Equatable, Sendable {
    let encoded: String
    let amount: UInt64?
    let unit: String?
    let description: String?
    let mints: [String]

    var isSatUnit: Bool {
        guard let unit else { return true }
        return unit.lowercased() == "sat"
    }
}

struct LightningRequestMetadata: Equatable, Sendable {
    let normalizedRequest: String
    let paymentMethod: PaymentMethodKind
    let amountSats: UInt64?
}

struct TokenPreview: Equatable, Sendable {
    let amount: UInt64
    let mintUrl: String
    let p2pkPubkeys: [String]
}

/// Non-main CDK boundary for synchronous FFI helpers.
///
/// CDK's wallet operations are mostly exposed as real async UniFFI futures, but
/// parsing helpers like `decodeInvoice`, `decodePaymentRequest`, and
/// `Token.decode` are synchronous. Keeping those behind this actor prevents
/// SwiftUI body recomputation and input handling from doing FFI work on the main
/// actor while still returning plain app values to views.
actor CdkRuntime {
    static let shared = CdkRuntime()

    private init() {}

    func decodePaymentRequest(
        _ raw: String,
        includeCashuPaymentRequests: Bool = false,
        preferCashuPaymentRequests: Bool = false
    ) -> PaymentRequestDecodeResult {
        PaymentRequestDecoder.decode(
            raw,
            includeCashuPaymentRequests: includeCashuPaymentRequests,
            preferCashuPaymentRequests: preferCashuPaymentRequests
        )
    }

    func normalizedLightningRequest(from raw: String) -> String? {
        PaymentRequestDecoder.encodedLightningRequest(from: raw)
    }

    func lightningMetadata(from raw: String) -> LightningRequestMetadata? {
        let normalized = PaymentRequestDecoder.encodedLightningRequest(from: raw)
            ?? PaymentRequestParser.normalizeLightningRequest(raw)

        guard !normalized.isEmpty,
              let decoded = try? decodeInvoice(invoiceStr: normalized) else {
            return nil
        }

        let paymentMethod: PaymentMethodKind
        switch decoded.paymentType {
        case .bolt11:
            paymentMethod = .bolt11
        case .bolt12:
            paymentMethod = .bolt12
        }

        return LightningRequestMetadata(
            normalizedRequest: normalized,
            paymentMethod: paymentMethod,
            amountSats: decoded.amountMsat.map(PaymentRequestDecoder.satsCeiling(from:))
        )
    }

    func tokenPreview(from tokenString: String) throws -> TokenPreview {
        let token = try Token.decode(encodedToken: tokenString)
        return TokenPreview(
            amount: try token.value().value,
            mintUrl: try token.mintUrl().url,
            p2pkPubkeys: token.p2pkPubkeys()
        )
    }
}

/// Typed result of decoding a raw payment request string.
enum PaymentRequestDecodeResult: Equatable, Sendable {
    case lightningAddress(String)
    case bolt11(amountSats: UInt64?, description: String?)
    case bolt12(amountSats: UInt64?, description: String?)
    case onchain(String)
    case cashuPaymentRequest(CashuPaymentRequestSummary)
    case unrecognized
}

extension PaymentRequestDecodeResult {
    /// Clean caution copy when this is a Lightning request that carries no amount,
    /// so every melt entry point surfaces identical, plain-language wording instead
    /// of a failing quote round-trip that leaks raw CDK codes. Nil for anything that
    /// isn't an amountless BOLT11 invoice or BOLT12 offer.
    var amountlessMeltCaution: String? {
        switch self {
        case .bolt11(let amountSats, _) where amountSats == nil:
            return "This invoice doesn't set an amount. Ask the sender for one with the amount set."
        case .bolt12(let amountSats, _) where amountSats == nil:
            return "This offer doesn't set an amount. Ask the sender for one with the amount set."
        default:
            return nil
        }
    }
}

enum PaymentRequestMode: String, Equatable, Sendable {
    case lightning
    case onchain
}

/// Centralized payment-request decoder. Wraps `PaymentRequestParser` +
/// CashuDevKit's `decodeInvoice` so the chip preview, recents tap, scan
/// callback, and live decode feedback all share a single classification path.
enum PaymentRequestDecoder {
    static func decode(
        _ raw: String,
        includeCashuPaymentRequests: Bool = false,
        preferCashuPaymentRequests: Bool = false
    ) -> PaymentRequestDecodeResult {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return .unrecognized }

        if includeCashuPaymentRequests,
           preferCashuPaymentRequests,
           let summary = cashuPaymentRequestSummary(from: trimmed) {
            return .cashuPaymentRequest(summary)
        }

        if let decoded = decodedLightningRequest(from: trimmed) {
            return decoded
        }

        if PaymentRequestParser.isHumanReadableLightningAddress(trimmed) {
            return .lightningAddress(trimmed)
        }

        if PaymentRequestParser.isBitcoinAddress(trimmed) {
            return .onchain(PaymentRequestParser.normalizeBitcoinRequest(trimmed))
        }

        if includeCashuPaymentRequests,
           let summary = cashuPaymentRequestSummary(from: trimmed) {
            return .cashuPaymentRequest(summary)
        }

        return .unrecognized
    }

    static func encodedLightningRequest(from raw: String) -> String? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        if let bitcoinURI = bitcoinPaymentURI(from: trimmed),
           let lightning = bitcoinURI.lightning {
            return PaymentRequestParser.normalizeLightningRequest(lightning)
        }

        let normalized = PaymentRequestParser.normalizeLightningRequest(trimmed)
        guard (try? decodeInvoice(invoiceStr: normalized)) != nil else {
            return nil
        }

        return normalized
    }

    static func encodedCashuPaymentRequest(from raw: String) -> String? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        if let bitcoinURI = bitcoinPaymentURI(from: trimmed),
           let creq = bitcoinURI.creq {
            return creq
        }

        let withoutCashuScheme = stripSchemePrefixes(["cashu://", "cashu:"], from: trimmed)
        let lowercased = withoutCashuScheme.lowercased()
        guard lowercased.hasPrefix("creqa") || lowercased.hasPrefix("creqb1") else {
            return nil
        }

        return withoutCashuScheme
    }

    static func parseCashuPaymentRequest(_ raw: String) throws -> Cdk.PaymentRequest {
        guard let encoded = encodedCashuPaymentRequest(from: raw) else {
            throw PaymentRequestDecodeError.noCashuPaymentRequest
        }

        return try decodePaymentRequest(encoded: encoded)
    }

    static func cashuPaymentRequestSummary(from raw: String) -> CashuPaymentRequestSummary? {
        guard let encoded = encodedCashuPaymentRequest(from: raw),
              let request = try? decodePaymentRequest(encoded: encoded) else {
            return nil
        }

        return CashuPaymentRequestSummary(
            encoded: encoded,
            amount: request.amount()?.value,
            unit: request.unit().map(unitDescription),
            description: request.description(),
            mints: request.mints()
        )
    }

    private static func decodedLightningRequest(from raw: String) -> PaymentRequestDecodeResult? {
        guard let normalized = encodedLightningRequest(from: raw),
              let decoded = try? decodeInvoice(invoiceStr: normalized) else {
            return nil
        }

        let amountSats: UInt64? = decoded.amountMsat.map(satsCeiling(from:))
        switch decoded.paymentType {
        case .bolt11:
            return .bolt11(amountSats: amountSats, description: decoded.description)
        case .bolt12:
            return .bolt12(amountSats: amountSats, description: decoded.description)
        }
    }

    fileprivate static func satsCeiling(from amountMsat: UInt64) -> UInt64 {
        amountMsat / 1000 + (amountMsat % 1000 == 0 ? 0 : 1)
    }

    /// True if the request carries an enforceable amount the user can't change
    /// (BOLT11 with amount, amountful BOLT12). Triggers auto-quote on tap.
    static func amountLocked(_ result: PaymentRequestDecodeResult) -> Bool {
        switch result {
        case .bolt11(let amount, _), .bolt12(let amount, _):
            return amount != nil
        case .lightningAddress, .onchain, .cashuPaymentRequest, .unrecognized:
            return false
        }
    }

    /// SF Symbol for the result type. Used by chip + live feedback.
    static func iconName(_ result: PaymentRequestDecodeResult) -> String {
        switch result {
        case .lightningAddress: return "at"
        case .bolt11, .bolt12: return "bolt.fill"
        case .onchain: return "bitcoinsign.circle"
        case .cashuPaymentRequest: return "banknote"
        case .unrecognized: return "questionmark.circle"
        }
    }

    /// Short human label for the type.
    static func typeLabel(_ result: PaymentRequestDecodeResult) -> String {
        switch result {
        case .lightningAddress: return "Lightning address"
        case .bolt11: return "BOLT11 invoice"
        case .bolt12: return "BOLT12 offer"
        case .onchain: return "Bitcoin address"
        case .cashuPaymentRequest: return "Cashu request"
        case .unrecognized: return "Unrecognized"
        }
    }

    /// `prefix(6)…suffix(6)` short representation for invoices and addresses;
    /// human-readable addresses are returned in full.
    static func shortRepresentation(_ raw: String, result: PaymentRequestDecodeResult) -> String {
        switch result {
        case .lightningAddress(let address):
            return address
        case .cashuPaymentRequest(let summary):
            return summary.description ?? amountLabel(for: summary) ?? "Cashu payment request"
        case .bolt11, .bolt12, .onchain, .unrecognized:
            let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
            guard trimmed.count > 16 else { return trimmed }
            return "\(trimmed.prefix(8))…\(trimmed.suffix(6))"
        }
    }

    static func amountLabel(for summary: CashuPaymentRequestSummary) -> String? {
        guard let amount = summary.amount else { return nil }
        return "\(amount) \(summary.unit ?? "sat")"
    }

    static func suggestedMode(_ result: PaymentRequestDecodeResult) -> MeltView.MeltMode? {
        switch result {
        case .bolt11, .bolt12, .lightningAddress:
            return .lightning
        case .onchain:
            return .onchain
        case .cashuPaymentRequest, .unrecognized:
            return nil
        }
    }

    static func unitDescription(_ unit: Cdk.CurrencyUnit) -> String {
        switch unit {
        case .sat:
            return "sat"
        case .msat:
            return "msat"
        case .usd:
            return "usd"
        case .eur:
            return "eur"
        case .auth:
            return "auth"
        case .custom(let unit):
            return unit
        }
    }

    private static func bitcoinPaymentURI(from raw: String) -> BitcoinPaymentURI? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let components = URLComponents(string: trimmed),
              components.scheme?.lowercased() == "bitcoin" else {
            return nil
        }

        let queryItems = components.queryItems ?? []
        let creq = queryValue(named: ["creq"], in: queryItems)
        let lightning = queryValue(named: ["lightning", "lightninginvoice"], in: queryItems)

        return BitcoinPaymentURI(creq: creq, lightning: lightning)
    }

    private static func queryValue(named names: Set<String>, in queryItems: [URLQueryItem]) -> String? {
        queryItems.first { names.contains($0.name.lowercased()) }?
            .value?
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func stripSchemePrefixes(_ prefixes: [String], from input: String) -> String {
        for prefix in prefixes where input.lowercased().hasPrefix(prefix) {
            return String(input.dropFirst(prefix.count))
        }
        return input
    }

    private struct BitcoinPaymentURI {
        let creq: String?
        let lightning: String?
    }

    private enum PaymentRequestDecodeError: Error {
        case noCashuPaymentRequest
    }
}

File: android/app/src/main/java/com/cashu/me/Core/PaymentRequestDecoder.kt
package com.cashu.me.Core

import com.cashu.me.Models.PaymentMethodKind
import org.cashudevkit.CurrencyUnit as CdkCurrencyUnit
import org.cashudevkit.PaymentType as CdkPaymentType
import org.cashudevkit.decodeInvoice
import org.cashudevkit.decodePaymentRequest
import java.net.URLDecoder
import java.nio.charset.StandardCharsets
import java.util.Base64

data class CashuPaymentRequestSummary(
    val encoded: String,
    val amount: Long? = null,
    val unit: String? = null,
    val description: String? = null,
    val mints: List<String> = emptyList(),
) {
    val isSatUnit: Boolean get() = unit?.lowercase() == null || unit.lowercase() == "sat"
}

sealed interface PaymentRequestDecodeResult {
    data class LightningAddress(val address: String) : PaymentRequestDecodeResult
    data class Bolt11(val amountSats: Long?, val description: String?) : PaymentRequestDecodeResult
    data class Bolt12(val amountSats: Long?, val description: String?) : PaymentRequestDecodeResult
    data class Onchain(val address: String) : PaymentRequestDecodeResult
    data class CashuPaymentRequest(val summary: CashuPaymentRequestSummary) : PaymentRequestDecodeResult
    data object Unrecognized : PaymentRequestDecodeResult
}

object PaymentRequestParser {
    fun normalizeLightningRequest(request: String): String {
        val trimmed = request.trim()
        return when {
            trimmed.startsWith("lightning://", ignoreCase = true) -> trimmed.drop("lightning://".length)
            trimmed.startsWith("lightning:", ignoreCase = true) -> trimmed.drop("lightning:".length)
            else -> trimmed
        }
    }

    fun normalizeBitcoinRequest(request: String): String {
        val trimmed = request.trim()
        val withoutScheme = when {
            trimmed.startsWith("bitcoin://", ignoreCase = true) -> trimmed.drop("bitcoin://".length)
            trimmed.startsWith("bitcoin:", ignoreCase = true) -> trimmed.drop("bitcoin:".length)
            else -> trimmed
        }
        return withoutScheme.substringBefore("?")
    }

    fun isBitcoinAddress(request: String): Boolean =
        BitcoinAddressValidator.isValidAddress(normalizeBitcoinRequest(request))

    fun isHumanReadableLightningAddress(request: String): Boolean {
        val trimmed = request.trim()
        val atIndex = trimmed.indexOf('@')
        if (atIndex <= 0 || atIndex == trimmed.lastIndex) return false
        val domain = trimmed.substring(atIndex + 1)
        return "." in domain && !domain.startsWith(".") && !domain.endsWith(".")
    }

    fun paymentMethod(request: String): PaymentMethodKind? {
        if (isHumanReadableLightningAddress(request)) return null
        val normalized = PaymentRequestDecoder.encodedLightningRequest(request) ?: normalizeLightningRequest(request)
        runCatching { decodeInvoice(normalized) }.getOrNull()?.let { decoded ->
            return when (decoded.paymentType) {
                CdkPaymentType.BOLT11 -> PaymentMethodKind.Bolt11
                CdkPaymentType.BOLT12 -> PaymentMethodKind.Bolt12
            }
        }
        if (isBitcoinAddress(request)) return PaymentMethodKind.Onchain
        return null
    }
}

object PaymentRequestDecoder {
    private const val CREQ_A_PREFIX = "creqA"

    fun decode(
        raw: String,
        includeCashuPaymentRequests: Boolean = false,
        preferCashuPaymentRequests: Boolean = false,
    ): PaymentRequestDecodeResult {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return PaymentRequestDecodeResult.Unrecognized

        if (includeCashuPaymentRequests && preferCashuPaymentRequests) {
            cashuPaymentRequestSummary(trimmed)?.let { return PaymentRequestDecodeResult.CashuPaymentRequest(it) }
        }

        decodedLightningRequest(trimmed)?.let { return it }

        if (PaymentRequestParser.isHumanReadableLightningAddress(trimmed)) {
            return PaymentRequestDecodeResult.LightningAddress(trimmed)
        }

        if (PaymentRequestParser.isBitcoinAddress(trimmed)) {
            return PaymentRequestDecodeResult.Onchain(PaymentRequestParser.normalizeBitcoinRequest(trimmed))
        }

        if (includeCashuPaymentRequests) {
            cashuPaymentRequestSummary(trimmed)?.let { return PaymentRequestDecodeResult.CashuPaymentRequest(it) }
        }

        return PaymentRequestDecodeResult.Unrecognized
    }

    fun encodedLightningRequest(raw: String): String? {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return null
        bitcoinPaymentURI(trimmed)?.lightning?.let { return PaymentRequestParser.normalizeLightningRequest(it) }
        val normalized = PaymentRequestParser.normalizeLightningRequest(trimmed)
        return if (runCatching { decodeInvoice(normalized) }.isSuccess) normalized else null
    }

    fun encodedCashuPaymentRequest(raw: String): String? {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return null
        bitcoinPaymentURI(trimmed)?.creq?.let { return it }
        val withoutCashuScheme = stripSchemePrefixes(listOf("cashu://", "cashu:"), trimmed)
        val lower = withoutCashuScheme.lowercase()
        return if (lower.startsWith("creqa") || lower.startsWith("creqb1")) withoutCashuScheme else null
    }

    fun cdkCompatibleCashuPaymentRequest(raw: String): String? {
        val encoded = encodedCashuPaymentRequest(raw) ?: return null
        return if (encoded.startsWith(CREQ_A_PREFIX, ignoreCase = true)) {
            val payload = encoded.drop(CREQ_A_PREFIX.length).trimEnd('=')
            val padding = (4 - payload.length % 4) % 4
            CREQ_A_PREFIX + payload + "=".repeat(padding)
        } else {
            encoded
        }
    }

    fun cashuPaymentRequestSummary(raw: String): CashuPaymentRequestSummary? {
        val encoded = encodedCashuPaymentRequest(raw) ?: return null
        val cdkEncoded = cdkCompatibleCashuPaymentRequest(encoded) ?: encoded
        runCatching { decodePaymentRequest(cdkEncoded) }.getOrNull()?.let { request ->
            return CashuPaymentRequestSummary(
                encoded = cdkEncoded,
                amount = request.amount()?.value?.toLong(),
                unit = request.unit()?.toDomainUnit(),
                description = request.description(),
                mints = request.mints(),
            )
        }
        return legacyCreqASummary(cdkEncoded)
    }

    private fun decodedLightningRequest(raw: String): PaymentRequestDecodeResult? {
        val normalized = encodedLightningRequest(raw) ?: return null
        val decoded = runCatching { decodeInvoice(normalized) }.getOrNull() ?: return null
        val amountSats = decoded.amountMsat?.let { (it.toLong() + 999) / 1000 }
        return when (decoded.paymentType) {
            CdkPaymentType.BOLT11 -> PaymentRequestDecodeResult.Bolt11(amountSats, decoded.description)
            CdkPaymentType.BOLT12 -> PaymentRequestDecodeResult.Bolt12(amountSats, decoded.description)
        }
    }

    fun amountLocked(result: PaymentRequestDecodeResult): Boolean = when (result) {
        is PaymentRequestDecodeResult.Bolt11 -> result.amountSats != null
        is PaymentRequestDecodeResult.Bolt12 -> result.amountSats != null
        else -> false
    }

    fun typeLabel(result: PaymentRequestDecodeResult): String = when (result) {
        is PaymentRequestDecodeResult.LightningAddress -> "Lightning address"
        is PaymentRequestDecodeResult.Bolt11 -> "BOLT11 invoice"
        is PaymentRequestDecodeResult.Bolt12 -> "BOLT12 offer"
        is PaymentRequestDecodeResult.Onchain -> "Bitcoin address"
        is PaymentRequestDecodeResult.CashuPaymentRequest -> "Cashu request"
        PaymentRequestDecodeResult.Unrecognized -> "Unrecognized"
    }

    fun shortRepresentation(raw: String, result: PaymentRequestDecodeResult): String = when (result) {
        is PaymentRequestDecodeResult.LightningAddress -> result.address
        is PaymentRequestDecodeResult.CashuPaymentRequest ->
            result.summary.description ?: amountLabel(result.summary) ?: "Cashu payment request"
        else -> {
            val trimmed = raw.trim()
            if (trimmed.length > 16) "${trimmed.take(8)}...${trimmed.takeLast(6)}" else trimmed
        }
    }

    fun amountLabel(summary: CashuPaymentRequestSummary): String? =
        summary.amount?.let { "$it ${summary.unit ?: "sat"}" }

    private fun bitcoinPaymentURI(raw: String): BitcoinPaymentURI? {
        val body = when {
            raw.startsWith("bitcoin://", ignoreCase = true) -> raw.drop("bitcoin://".length)
            raw.startsWith("bitcoin:", ignoreCase = true) -> raw.drop("bitcoin:".length)
            else -> return null
        }
        val query = body.substringAfter("?", missingDelimiterValue = "")
        if (query.isEmpty()) return BitcoinPaymentURI(creq = null, lightning = null)
        val parameters = query.split("&").mapNotNull { entry ->
            val key = entry.substringBefore("=", missingDelimiterValue = "").decodeQueryComponent().trim()
            if (key.isEmpty()) return@mapNotNull null
            val value = entry.substringAfter("=", missingDelimiterValue = "").decodeQueryComponent().trim()
            key to value
        }
        val creq = parameters
            .firstOrNull { (key, _) -> key.equals("creq", ignoreCase = true) }
            ?.second
        val lightning = parameters
            .firstOrNull { (key, _) -> key.equals("lightning", ignoreCase = true) || key.equals("lightninginvoice", ignoreCase = true) }
            ?.second
        return BitcoinPaymentURI(creq = creq, lightning = lightning)
    }

    private fun stripSchemePrefixes(prefixes: List<String>, input: String): String {
        val prefix = prefixes.firstOrNull { input.startsWith(it, ignoreCase = true) }
        return if (prefix == null) input else input.drop(prefix.length)
    }

    private fun legacyCreqASummary(encoded: String): CashuPaymentRequestSummary? {
        if (!encoded.startsWith(CREQ_A_PREFIX, ignoreCase = true)) return null
        val payload = encoded.drop(CREQ_A_PREFIX.length)
        val bytes = runCatching { Base64.getUrlDecoder().decode(payload) }.getOrNull() ?: return null
        val fields = (runCatching { CborReader(bytes).readValue() }.getOrNull() as? CborValue.Map)
            ?.entries
            ?.mapNotNull { (key, value) -> (key as? CborValue.Text)?.value?.let { it to value } }
            ?.toMap()
            ?: return null
        val mints = (fields["m"] as? CborValue.Array)
            ?.values
            ?.mapNotNull { (it as? CborValue.Text)?.value }
            .orEmpty()
        return CashuPaymentRequestSummary(
            encoded = encoded,
            amount = (fields["a"] as? CborValue.UInt)?.value,
            unit = (fields["u"] as? CborValue.Text)?.value,
            description = (fields["d"] as? CborValue.Text)?.value,
            mints = mints,
        )
    }

    private data class BitcoinPaymentURI(val creq: String?, val lightning: String?)

    private sealed interface CborValue {
        data class UInt(val value: Long) : CborValue
        data class Text(val value: String) : CborValue
        data class Bool(val value: Boolean) : CborValue
        data class Array(val values: List<CborValue>) : CborValue
        data class Map(val entries: List<Pair<CborValue, CborValue>>) : CborValue
        data object Null : CborValue
    }

    private class CborReader(private val bytes: ByteArray) {
        private var position = 0

        fun readValue(): CborValue {
            val initial = readByte()
            val major = initial shr 5
            val additional = initial and 0x1F
            return when (major) {
                0 -> CborValue.UInt(readLength(additional))
                3 -> CborValue.Text(readBytes(readLength(additional).toInt()).toString(Charsets.UTF_8))
                4 -> CborValue.Array(List(readLength(additional).toInt()) { readValue() })
                5 -> CborValue.Map(List(readLength(additional).toInt()) { readValue() to readValue() })
                7 -> when (additional) {
                    20 -> CborValue.Bool(false)
                    21 -> CborValue.Bool(true)
                    22 -> CborValue.Null
                    else -> error("Unsupported CBOR simple value.")
                }
                else -> error("Unsupported CBOR major type.")
            }
        }

        private fun readLength(additional: Int): Long = when {
            additional < 24 -> additional.toLong()
            additional == 24 -> readByte().toLong()
            additional == 25 -> ((readByte() shl 8) or readByte()).toLong()
            additional == 26 -> (0 until 4).fold(0L) { acc, _ -> (acc shl 8) or readByte().toLong() }
            additional == 27 -> (0 until 8).fold(0L) { acc, _ -> (acc shl 8) or readByte().toLong() }
            else -> error("Unsupported CBOR length.")
        }

        private fun readBytes(length: Int): ByteArray {
            require(length >= 0 && position + length <= bytes.size) { "Truncated CBOR payload." }
            return bytes.copyOfRange(position, position + length).also { position += length }
        }

        private fun readByte(): Int {
            require(position < bytes.size) { "Truncated CBOR payload." }
            return bytes[position++].toInt() and 0xFF
        }
    }

    private fun String.decodeQueryComponent(): String =
        runCatching { URLDecoder.decode(this, StandardCharsets.UTF_8.name()) }.getOrDefault(this)

    private fun CdkCurrencyUnit.toDomainUnit(): String = when (this) {
        CdkCurrencyUnit.Sat -> "sat"
        CdkCurrencyUnit.Msat -> "msat"
        CdkCurrencyUnit.Usd -> "usd"
        CdkCurrencyUnit.Eur -> "eur"
        CdkCurrencyUnit.Auth -> "auth"
        is CdkCurrencyUnit.Custom -> unit
    }
}
